dttools/src/catalog_query.c

```diff
@@ -98,53 +120,43 @@ struct catalog_query *catalog_query_create(const char *hosts, struct jx *filter_
 		}
 	} while (next_host);
 
-	sorted_hosts = list_splice(previously_up, previously_down);
+	return list_splice(previously_up, previously_down);
+}
+
+struct catalog_query *catalog_query_create(const char *hosts, struct jx *filter_expr, time_t stoptime)
+{
+	struct catalog_query *q = NULL;
+	char *n;
+	struct catalog_host *h;
+	struct list *sorted_hosts = catalog_query_sort_hostlist(hosts);
 
 	list_first_item(sorted_hosts);
 	while((h = list_next_item(sorted_hosts))) {
-		struct link *link = http_query(h->url, "GET", stoptime);
-
-		if(!link) {
-			goto FAILURE;
-		}
-
-		struct jx *j = jx_parse_link(link,stoptime);
-
-		link_close(link);
-
-		if(!j) {
-			debug(D_DEBUG,"query result failed to parse as JSON");
-			goto FAILURE;
-		}
-
-		if(!jx_istype(j,JX_ARRAY)) {
-			debug(D_DEBUG,"query result is not a JSON array");
-			jx_delete(j);
-			goto FAILURE;
-		}
-
-		q = xxmalloc(sizeof(*q));
-		q->data = j;
-		q->current = j->u.items;
-		q->filter_expr = filter_expr;
-
-		if(h->down) {
-			debug(D_DEBUG,"catalog server at %s is back up", h->host);
-			set_first_element(down_hosts);
-			while((n = set_next_element(down_hosts))) {
-				if(!strcmp(n, h->host)) {
-					free(n);
-					set_remove(down_hosts, n);
-					break;
+		struct jx *j = catalog_query_send_query(h->url, stoptime);
```

Comment:
That seems like a pretty reasonable solution. I can add a check to respect `stoptime` and just keep going around the list until then